src/sr_single_entity_handler.erl

```diff
@@ -0,0 +1,121 @@
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%%% Cowboy Callbacks
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+-spec rest_init(cowboy_req:req(), options()) ->
+  {ok, cowboy_req:req(), state()}.
+rest_init(Req, Opts) ->
+  Req1 = announce_req(Req, Opts),
+  {Id, Req2} = cowboy_req:binding(id, Req1),
+  {ok, Req2, #{opts => Opts, id => Id}}.
+
+-spec resource_exists(cowboy_req:req(), state()) ->
+  {boolean(), cowboy_req:req(), state()}.
+resource_exists(Req, State) ->
+  #{opts := #{model := Model}, id := Id} = State,
+  case sumo:find(Model, Id) of
+    notfound -> {false, Req, State};
+    Entity -> {true, Req, State#{entity => Entity}}
+  end.
+
+%% @todo Use swagger's 'consumes' to auto-generate this if possible
+%% @see https://github.com/inaka/sumo_rest/issues/7
+-spec content_types_accepted(cowboy_req:req(), state()) ->
+  {[{{binary(), binary(), '*'}, atom()}], cowboy_req:req(), state()}.
+content_types_accepted(Req, State) ->
+  {[{{<<"application">>, <<"json">>, '*'}, handle_put}], Req, State}.
+
+-spec handle_get(cowboy_req:req(), state()) ->
+  {iodata(), cowboy_req:req(), state()}.
+handle_get(Req, State) ->
+  #{opts := #{model := Model}, entity := Entity} = State,
+  ResBody = sr_json:encode(Model:to_json(Entity)),
+  {ResBody, Req, State}.
+
+-spec handle_put(cowboy_req:req(), state()) ->
+  {{true, binary()} | false | halt, cowboy_req:req(), state()}.
+handle_put(Req, #{entity := Entity} = State) ->
+  #{opts := #{model := Model}} = State,
+  try
+    {ok, Body, Req1} = cowboy_req:body(Req),
+    Json             = sr_json:decode(Body),
+    handle_put(Model:update(Entity, Json), Req1, State)
+  catch
+    _:badjson ->
+      Req3 = cowboy_req:set_resp_body(<<"Malformed JSON request">>, Req),
+      {false, Req3, State};
+    _:Exception -> handle_exception(Exception, Req, State)
+  end;
+handle_put(Req, #{id := Id} = State) ->
+  #{opts := #{model := Model}} = State,
+  try
+    {ok, Body, Req1} = cowboy_req:body(Req),
+    Json             = sr_json:decode(Body),
+    handle_put(from_json(Model, Id, Json), Req1, State)
+  catch
+    _:badjson ->
+      Req3 = cowboy_req:set_resp_body(<<"Malformed JSON request">>, Req),
+      {false, Req3, State};
+    _:Exception -> handle_exception(Exception, Req, State)
+  end.
+
+-spec delete_resource(cowboy_req:req(), state()) ->
+  {boolean() | halt, cowboy_req:req(), state()}.
+delete_resource(Req, State) ->
+  #{opts := #{model := Model}, id := Id} = State,
+  try
+    Result = sumo:delete(Model, Id),
+    {Result, Req, State}
+  catch
+    _:Exception -> handle_exception(Exception, Req, State)
+  end.
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%%% Auxiliary Functions
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+from_json(Model, Id, Json) ->
+  case erlang:function_exported(Model, from_json, 2) of
```

Comment:
The "problem" with incredibly useful and generic code that does things automagically is that it is often hard to debug when things go wrong. E.g. if the user programmer fails to export the proper functions, all they will get is a stacktrace. If exporting a certain function is a requirement and a source of a possibly foreseeable error, should it not be handled?